android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt

```diff
@@ -2,55 +2,85 @@
+  private val activeNetworkLock = Any()
+
   // requestNetwork attempts to find the best network that matches the passed NetworkRequest. It is
   // possible that this might return an unusuable network, eg a captive portal.
   fun monitorDnsChanges(connectivityManager: ConnectivityManager, dns: DnsConfig) {
-    val networkConnectivityRequest =
-        NetworkRequest.Builder()
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
-            .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
-            .build()
-
-    connectivityManager.registerNetworkCallback(
-        networkConnectivityRequest,
+
+    connectivityManager.registerDefaultNetworkCallback(
         object : ConnectivityManager.NetworkCallback() {
           override fun onAvailable(network: Network) {
-            super.onAvailable(network)
+            val networkCapabilities = connectivityManager.getNetworkCapabilities(network)
+            if (networkCapabilities != null &&
+                networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN) &&
+                networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
+              // Only use if the network in the callback is the current active network. Network
+              // updates are not guaranteed to come in order.
+              synchronized(activeNetworkLock) {
+                if (connectivityManager.activeNetwork == network) {
+                  activeNetwork = network
+                  super.onAvailable(network)
 
-            val sb = StringBuilder()
-            val linkProperties: LinkProperties? = connectivityManager.getLinkProperties(network)
-            val dnsList: MutableList<InetAddress> = linkProperties?.dnsServers ?: mutableListOf()
-            for (ip in dnsList) {
-              sb.append(ip.hostAddress).append(" ")
-            }
-            val searchDomains: String? = linkProperties?.domains
-            if (searchDomains != null) {
-              sb.append("\n")
-              sb.append(searchDomains)
-            }
+                  val sb = StringBuilder()
+                  val linkProperties: LinkProperties? =
+                      connectivityManager.getLinkProperties(network)
+                  val dnsList: MutableList<InetAddress> =
+                      linkProperties?.dnsServers ?: mutableListOf()
+                  for (ip in dnsList) {
+                    sb.append(ip.hostAddress).append(" ")
+                  }
+                  val searchDomains: String? = linkProperties?.domains
+                  if (searchDomains != null) {
+                    sb.append("\n")
+                    sb.append(searchDomains)
+                  }
+
+                  if (dns.updateDNSFromNetwork(sb.toString())) {
+                    Libtailscale.onDNSConfigChanged(linkProperties?.interfaceName)
+                  } else {
+                    Log.d(TAG, "DNS config not updated")
```

Comment:
```suggestion
                    Log.d(TAG, "DNS config not updated because active network $connectivityManager.activeNetwork.toString() does not match network $network")
```